core/src/main/java/org/elasticsearch/index/query/TermsQueryParser.java

```diff
@@ -159,58 +100,69 @@ public Query parse(QueryShardContext context) throws IOException, QueryParsingEx
         }
 
         if (fieldName == null) {
-            throw new QueryParsingException(parseContext, "terms query requires a field name, followed by array of terms");
+            throw new QueryParsingException(parseContext, "terms query requires a field name, followed by array of terms or a document lookup specification");
         }
-
-        MappedFieldType fieldType = context.fieldMapper(fieldName);
-        if (fieldType != null) {
-            fieldName = fieldType.names().indexName();
+        TermsQueryBuilder termsQueryBuilder;
+        if (values == null) {
+            termsQueryBuilder = new TermsQueryBuilder(fieldName);
+        } else {
+            termsQueryBuilder = new TermsQueryBuilder(fieldName, values);
```

Comment:
looking deeper, I see that we set a non null TermsLookup object only when we have it in query, which causes a validation error when values are set too. We should keep it that way then, this is as good as it gets.